src/twisted/protocols/_smb/core.py

```diff
@@ -0,0 +1,760 @@
+                    | FILE_DELETE_CHILD
+                    | FILE_EXECUTE
+                    | FILE_READ_ATTRIBUTES
+                    | FILE_WRITE_ATTRIBUTES
+                    | DELETE
+                    | READ_CONTROL
+                    | WRITE_DAC
+                    | WRITE_OWNER
+                    | SYNCHRONIZE
+                ),
+            )
+        if IPipe.providedBy(share):
+            assert resp is None, "share can only be one type"
+            resp = resp_type(
+                share_type=SHARE_PIPE,
+                flags=0,
+                max_perms=(
+                    FILE_READ_DATA
+                    | FILE_WRITE_DATA
+                    | FILE_APPEND_DATA
+                    | FILE_READ_EA
+                    |
+                    # FILE_WRITE_EA |
+                    # FILE_DELETE_CHILD |
+                    FILE_EXECUTE
+                    | FILE_READ_ATTRIBUTES
+                    |
+                    # FILE_WRITE_ATTRIBUTES |
+                    DELETE
+                    | READ_CONTROL
+                    |
+                    # WRITE_DAC |
+                    # WRITE_OWNER |
+                    SYNCHRONIZE
+                ),
+            )
+        if IPrinter.providedBy(share):
+            assert resp is None, "share can only be one type"
+            resp = resp_type(
+                share_type=SHARE_PRINTER,
+                flags=0,
+                # FIXME need to check printer  max perms
+                max_perms=(
+                    FILE_READ_DATA
+                    | FILE_WRITE_DATA
+                    | FILE_APPEND_DATA
+                    | FILE_READ_EA
+                    |
+                    # FILE_WRITE_EA |
+                    # FILE_DELETE_CHILD |
+                    FILE_EXECUTE
+                    | FILE_READ_ATTRIBUTES
+                    |
+                    # FILE_WRITE_ATTRIBUTES |
+                    DELETE
+                    | READ_CONTROL
+                    |
+                    # WRITE_DAC |
+                    # WRITE_OWNER |
+                    SYNCHRONIZE
+                ),
+            )
+        if resp is None:
+            log.error("unknown share object {share!r}", share=share)
+            errorResponse(packet, NTStatus.UNSUCCESSFUL)
+            return
+        packet.hdr.tree_id = _base.int32key(packet.ctx["trees"], share)
+        packet.data = _base.pack(resp)
+        sendHeader(packet)
+
+    d.addCallback(cb_tree)
+    d.addErrback(eb_tree)
+
+
+def smb_tree_disconnect(packet: _base.SMBPacket, resp_type: type) -> None:
+    del packet.ctx["trees"][packet.hdr.tree_id]
+    packet.data = _base.pack(resp_type())
+    sendHeader(packet)
+
+
+class SMBFactory(protocol.Factory):
+    """
+    Factory for SMB servers
+    """
+
+    def __init__(self, portal: Portal, domain: str = "WORKGROUP") -> None:
+        """
+        @param portal: the configured portal
+        @type portal: L{twisted.cred.portal.Portal}
```

Comment:
Maybe say something about the expectation of the portal's configured checkers & realm?